File: Python/BOJ/1654 랜선 자르기.py
def search(arr):  # Parametric Search
    start = 1
    end = max(x)
    while start <= end:
        middle = (start + end) // 2
        num = 0
        for i in x:
            num += i//middle
        if num >= N:
            start = middle + 1
        else:
            end = middle - 1
    return (end)


K, N = map(int, input().split())
x = [0]*K
for w in range(K):
    x[w] = int(input())
print(search(x))

File: Java/BOJ/1654 랜선 자르기.java
import java.io.*;
import java.util.*;
import java.math.*;

public class Main {
	static long[] x;
	static int K;
	static int N;
	static long search() {
		long start = 1;
		long end = x[K-1];
		while (start <= end) {
			long mid = (start + end)/2;
			long num = 0;
			for (int i = 0; i< K; i++) {
				num += x[i]/mid;
			}
			if (num >= N) {
				start = mid + 1;
			}
			else {
				end = mid - 1;
			}
		}
		return end;
	}
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        K = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());
        x = new long[K];
        for (int i = 0 ; i < K; i++) {
        	x[i] = Long.parseLong(br.readLine());
        }
        Arrays.sort(x);;
        System.out.println(search());
    }
}